services/rpz.c

```diff
@@ -938,107 +1293,692 @@ log_rpz_apply(uint8_t* dname, enum rpz_action a, struct query_info* qinfo,
+		ret = NULL;
+	}
+
+	return ret;
+}
+
+static struct local_zone*
+rpz_delegation_point_zone_lookup(struct delegpt* dp, struct local_zones* zones,
+	uint16_t qclass,
+	/* output parameter */
+	struct matched_delegation_point* match)
+{
+	struct delegpt_ns* nameserver;
+	struct local_zone* z = NULL;
+
+	rpz_log_dname("delegation point", dp->name, dp->namelen);
+	// XXX: do we want this?
+	z = rpz_find_zone(zones, dp->name, dp->namelen, qclass, 0, 0, 0);
+	if(z != NULL) {
+		match->dname = dp->name;
+		match->dname_len = dp->namelen;
+	} else if(z == NULL) {
+		for(nameserver = dp->nslist;
+		    nameserver != NULL;
+		    nameserver = nameserver->next) {
+			rpz_log_dname("delegation point", nameserver->name, nameserver->namelen);
+			z = rpz_find_zone(zones, nameserver->name, nameserver->namelen,
+					  qclass, 0, 0, 0);
+			if(z != NULL) {
+				match->dname = nameserver->name;
+				match->dname_len = nameserver->namelen;
+				break;
+			}
+		}
+	}
+
+	verbose(VERB_ALGO, "rpz: delegation point zone found=%d", z != NULL);
+
+	return z;
+}
+
+struct dns_msg*
+rpz_callback_from_iterator_module(struct module_qstate* ms, struct iter_qstate* is)
+{
+	struct auth_zones* az;
+	struct auth_zone* a;
+	struct clientip_synthesized_rr* raddr = NULL;
+	struct rpz* r;
+	struct local_zone* z = NULL;
+	struct matched_delegation_point match = {0};
+
+	if(ms->env == NULL || ms->env->auth_zones == NULL) { return 0; }
+
+	az = ms->env->auth_zones;
+
+	verbose(VERB_ALGO, "rpz: iterator module callback: have_rpz=%d", az->rpz_first != NULL);
+
+	lock_rw_rdlock(&az->rpz_lock);
+
+	for(a = az->rpz_first; a != NULL; a = a->rpz_az_next) {
+		lock_rw_rdlock(&a->lock);
+		r = a->rpz;
+
+		// XXX: check rfc which action has preference
+
+		raddr = rpz_delegation_point_ipbased_trigger_lookup(r, is);
+		if(raddr != NULL) {
+			lock_rw_unlock(&a->lock);
+			break;
+		}
+
+		z = rpz_delegation_point_zone_lookup(is->dp, r->nsdname_zones,
+						     ms->qinfo.qclass, &match);
+		if(z != NULL) {
+			lock_rw_unlock(&a->lock);
+			break;
+		}
+		lock_rw_unlock(&a->lock);
+	}
+
+	lock_rw_unlock(&az->rpz_lock);
+
+	if(raddr == NULL && z == NULL) { return NULL; }
+	else if(raddr != NULL) { return rpz_apply_nsip_trigger(ms, r, raddr); }
+	else if(z != NULL) { return rpz_apply_nsdname_trigger(ms, r, z, &match); }
+	else { return NULL; }
+}
+
+static int
+rpz_apply_maybe_clientip_trigger(struct auth_zones* az, struct module_env* env,
+	struct query_info* qinfo, struct edns_data* edns, struct comm_reply* repinfo,
+	uint8_t* taglist, size_t taglen, struct ub_server_stats* stats,
+	sldns_buffer* buf, struct regional* temp,
+	/* output parameters */
+	struct local_zone** z_out, struct auth_zone** a_out, struct rpz** r_out)
+{
+	int ret = 0;
+	enum rpz_action client_action;
+	struct clientip_synthesized_rr* node = rpz_resolve_client_action_and_zone(
+		az, qinfo, repinfo, taglist, taglen, stats, z_out, a_out, r_out);
+
+	client_action = node == NULL ? RPZ_INVALID_ACTION : node->action;
```

Comment:
I would suggest parentheses around the if.  `(node == ... action);`  I worry that `=` is assigned `(node)` or `(node==NULL)`.

Reply:
Was fixed, closing it.